proposals/0066-feature-gate-threshold-automation.md

```diff
@@ -0,0 +1,245 @@
+---
+simd: '0066'
+title: Feature Gate Threshold Automation
+authors:
+  - Tyera Eulburg
+  - Joe Caulfield
+category: Standard
+type: Core
+status: Draft
+created: 2023-10-11
+feature: (fill in with feature tracking issues once accepted)
+---
+
+## Summary
+
+This SIMD outlines a proposal for automating the feature activation process
+based on a stake-weighted support threshold, rather than manual human action.
+
+## Motivation
+
+Feature gates wrap new cluster functionality, and typically change the rules of
+consensus. As such, a feature gate needs to be supported by a strong majority
+of cluster stake when it is activated, or else it risks partitioning the
+network. The current feature-gate system comprises two steps:
+
+1. An individual key-holder queues a feature gate for activation
+2. The runtime automatically activates the feature on the next epoch boundary
+
+The key-holder is the one who (with the help of the solana-cli) assesses the
+amount of stake that recognizes the feature and decides whether it is safe to
+activate. This is obviously brittle and subject to human error.
+
+This SIMD proposes that the runtime itself replaces the human to assess the
+amount of stake that supports a particular queued feature and only
+automatically activates it when a preset threshold is met. 
+
+## New Terminology
+
+- Feature Gate program: The BPF program that will own all feature accounts and
+possess the capability to revoke them.
+- Feature-Request account: The PDA under the Feature Gate program used to track
+features requested for activation.
+- Feature-Queue account: The PDA under the Feature Gate program used to track
+features queued for activation that must have support signaled by nodes.
+
+## Detailed Design
+
+Before this SIMD is published, we will deploy a BPF program to manage feature
+activations, which will allow us to revoke pending feature activations
+(previously not possible).
+The feature activation process will otherwise be unchanged, so no SIMD is
+required.
+This program will be upgradeable; the authority and process through which it
+will be upgraded is outside the scope of this SIMD. Since this SIMD requires an
+update to the program, the upgrade details will be determined in parallel.
+
+The initial Feature Gate Program can be found in this pull request to the
+Solana Program Library:
+<https://github.com/solana-labs/solana-program-library/pull/4987>.
+The runtime changes and feature activation that will enable this program can be
+found in this feature tracking issue:
+<https://github.com/solana-labs/solana/issues/33547>.
```

Comment:
I hate to say it but we probably should do a SIMD for this, as a prereq for this SIMD. Should be short and quick to do hopefully!
It feels like we kinda are just sneaking this functionally in here, where it really deserves to be stand-alone. Especially since we're allowing the program to be upgradable. Would be good, in the other SIMD, to describe who can upgrade it and for how long!

Reply:
Yeah, I agree, it seems like kind of a big deal to not have a SIMD for it.

I can write this SIMD up.

Reply:
Would you also suggest we delay activation of the [no-op swap feature gate](https://github.com/solana-labs/solana/issues/33547) pending results of this new SIMD?

Reply:
I guess we could just put the no-op there and then let the new SIMD determine the program stuff, but the authority aspect of this new BPF program would also apply to the no-op